table/merging_iterator.cc

```diff
@@ -343,86 +562,554 @@ class MergingIterator : public InternalIterator {
 
   IteratorWrapper* CurrentForward() const {
     assert(direction_ == kForward);
-    return !minHeap_.empty() ? minHeap_.top() : nullptr;
+    assert(minHeap_.empty() || minHeap_.top()->type == HeapItem::ITERATOR);
+    return !minHeap_.empty() ? &minHeap_.top()->iter : nullptr;
   }
 
   IteratorWrapper* CurrentReverse() const {
     assert(direction_ == kReverse);
     assert(maxHeap_);
-    return !maxHeap_->empty() ? maxHeap_->top() : nullptr;
+    assert(maxHeap_->empty() || maxHeap_->top()->type == HeapItem::ITERATOR);
+    return !maxHeap_->empty() ? &maxHeap_->top()->iter : nullptr;
   }
 };
 
-void MergingIterator::AddToMinHeapOrCheckStatus(IteratorWrapper* child) {
-  if (child->Valid()) {
-    assert(child->status().ok());
+// Seek to fist key >= target key (internal key) for children_[starting_level:].
+// Cascading seek optimizations are applied if range tombstones are present (see
+// comment above Seek() for more).
+//
+// @param range_tombstone_reseek Whether target is some range tombstone
+// end, i.e., whether this SeekImpl() call is a part of a "cascading seek". This
+// is used only for recoding relevant perf_context.
+void MergingIterator::SeekImpl(const Slice& target, size_t starting_level,
+                               bool range_tombstone_reseek) {
+  ClearHeaps();
+  status_ = Status::OK();
+  IterKey current_search_key;
+  current_search_key.SetInternalKey(target, false /* copy */);
+  // Seek target might change to some range tombstone end key, so
+  // we need to remember them for async requests.
+  // (level, target) pairs
+  autovector<std::pair<size_t, std::string>> prefetched_target;
+
+  for (auto level = starting_level; level < children_.size(); ++level) {
+    {
+      PERF_TIMER_GUARD(seek_child_seek_time);
+      children_[level].iter.Seek(current_search_key.GetInternalKey());
+    }
+
+    PERF_COUNTER_ADD(seek_child_seek_count, 1);
+
+    if (!range_tombstone_iters_.empty()) {
+      if (range_tombstone_reseek) {
+        // This seek is to some range tombstone end key.
+        // Should only happen when there are range tombstones.
+        PERF_COUNTER_ADD(internal_range_del_reseek_count, 1);
+      }
+      if (children_[level].iter.status().IsTryAgain()) {
+        prefetched_target.emplace_back(
+            level, current_search_key.GetInternalKey().ToString());
+      }
+      auto range_tombstone_iter = range_tombstone_iters_[level];
+      if (range_tombstone_iter) {
+        range_tombstone_iter->Seek(current_search_key.GetUserKey());
+        if (range_tombstone_iter->Valid()) {
+          InsertRangeTombstoneToMinHeap(level);
```

Comment:
> Does this add keys to the min-heap that are smaller than the seek key?

Yes.

>  or can we insert only the nearest endpoint at or larger than the seek key?

This should work, let me give it a try. The range tombstone start keys that are too small will be filtered out in `PopDeleteRangeStart()` calls, but I think this method of inserting the "right" key seems cleaner, and could be more efficient.

Reply:
> Do we need to add both start and end endpoints here
After giving it more thought, maybe we can always just have one end point of a range tombstone in the heap. When the start key is popped, we then add the end key of a range tombstone.

Reply:
> can we insert only the nearest endpoint at or larger than the seek key?

I don't think we can do this, consider the following example with point keys and range tombstones:

L1: [2, 9), 6
L2: 3, [7, 15)

3 is at top of the minHeap_, and we see [2, 9) covers it, so we call `SeekImpl(L2, 9)`. The range tombstone `[7, 15)` has 7 before seek key 9. We cannot simply add its end key 15 to the minHeap_, since the overall "current key" after the seek is key 6, which is before the *start key 7* of this range tombstone.

Reply:
Ignore the above. In this case, we can treat [7, 15) from L2 as active, as it only covers keys from L2+. It does make the meaning of "active" a little more complicated tho.